staging/src/k8s.io/client-go/tools/cache/reflector.go

```diff
@@ -318,6 +332,7 @@ func (r *Reflector) ListAndWatch(stopCh <-chan struct{}) error {
 
 		w, err := r.listerWatcher.Watch(options)
 		if err != nil {
```

Comment:
What about removing all the specific error handling here to a DefaultDropWatchHandler that is registered by default?
Because I'd expect people would like to handle EOF and ConnectionRefused error (if only for monitoring purposes).

Reply:
The most important thing for us is to have some mechanism to bubble up to user "hey, your auth tokens expired, and we can't talk to kubernetes anymore".

I'm deeply worried that this PR is going to turn into a quagmire where we can't add any error-handling mechanism at all until we've had a long discussion on which errors bubble up and which do not. 

I think that's a good discussion to have. But I don't feel like I'm well-equipped to facilitate that discussion, or to weigh competing needs (e.g., people who want the informer to retry EOF internally vs those who do not)

Reply:
Understood, but I'm just suggesting to keep the existing error handling but move it to a DefaultDropWatchHandler. That way a user can customize it without changing the existing default behaviour.

Reply:
hmmm...i spent some time playing around with this, and couldn't come up with a solution that fit well. This error handling depends on unexported fields and functions of the package (r.expectedTypeName, isExpiredError), and wasn't sure how much we should really be exposing.

Did you have a particular API in mind?

Reply:
@nicks A DefaultDropWatchHandler could use these fields, right? Something like this:
```
func DefaultDropWatchHandler(err error, doneCh <-chan struct{}) {
	switch {
			case isExpiredError(err):
				r.setIsLastSyncResourceVersionExpired(true)
				klog.V(4).Infof("%s: watch of %v closed with: %v", r.name, r.expectedTypeName, err)
			case err == io.EOF:
				// watch closed normally
			case err == io.ErrUnexpectedEOF:
				klog.V(1).Infof("%s: Watch for %v closed with unexpected EOF: %v", r.name, r.expectedTypeName, err)
			default:
				utilruntime.HandleError(fmt.Errorf("%s: Failed to watch %v: %v", r.name, r.expectedTypeName, err))
	}
	if !utilnet.IsConnectionRefused(err) {
		doneCh <- <- struct{}{}
	}
}
```

Reply:
ok, i pushed a new branch to demonstrate what that would look like, what do you think?